### PR TITLE
AE-471 UMA disputer

### PR DESCRIPTION
**Pull request by arbitrary-cr0**
Forta agent port of UMA's dispute bot: (https://github.com/UMAprotocol/protocol/blob/master/packages/disputer/src/disputer.js)

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -34,6 +35,12 @@ the following handlers:
   - Metadata field contains VotingToken contract address, caller address, and transaction hash
   - Must have traces enabled for this to operate correctly
 
+- AE-UMA-DISPUTE
+  - Fired when a liquidation can be disputed
+  - Severity is always "medium"
+  - Type is always set to "suspicious"
```
Comment (arbitrary-kdawg) [inline README.md:41]: Should type be "info" instead of "suspicious"?

File: src/bot-handlers/disputer.js [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+const {
+  umaEverestId,
+} = require('../../agent-config.json');
+
+const {
+  initializeContracts,
+} = require('./initialization');
+
+// web3 helpers
+const {
```
Comment (arbitrary-kdawg) [inline src/bot-handlers/disputer.js:13]: nit: make this a single line

File: src/bot-handlers/disputer.js [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+  toWei,
+  toBN,
+} = web3.utils;
+
+function createAlert(price, scaledPrice, liquidation) {
+  return Finding.fromObject({
+    name: 'Dispute Opportunity',
```
Comment (arbitrary-kdawg) [inline src/bot-handlers/disputer.js:20]: Add UMA to the name

File: src/bot-handlers/disputer.js [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+    description: 'Dispute opportunity identified',
```
Comment (arbitrary-kdawg) [inline src/bot-handlers/disputer.js:21]: Add the contract address (and maybe also the price) to the description?

File: src/bot-handlers/disputer.js [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+    alertId: 'AE-UMA-DISPUTE',
+    protocol: 'uma',
+    severity: FindingSeverity.Medium,
+    type: FindingType.Suspicious, // TODO: Change to info when we update forta-agent
```
Comment (arbitrary-kdawg) [inline src/bot-handlers/disputer.js:25]: TODO reminder

File: src/bot-handlers/disputer.js [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+      // eslint-disable-next-line no-unused-vars
+      const disputableLiquidationsWithPrices = (
+        await Promise.all(
+          undisputedLiquidations.map(async (liquidation) => {
+            const liquidationTime = parseInt(liquidation.liquidationTime.toString(), 10);
+            const lookback = priceFeed.getLookback();
+            const historicalLookbackWindow = priceFeed.getLastUpdateTime() - lookback;
+
+            if (liquidationTime < historicalLookbackWindow) {
+              // console.error('Liquidation time before earliest price feed historical timestamp');
```
Comment (arbitrary-kdawg) [inline src/bot-handlers/disputer.js:81]: what's the reason to keep the console.error() statements commented out?

File: src/bot-handlers/disputer.js [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+              return null;
+            }
+            // get the historic price at the liquidation time.
+            let price;
+            try {
+              price = await priceFeed.getHistoricalPrice(liquidationTime);
+            } catch (error) {
+              // console.error('could not get historical price');
```
Comment (arbitrary-kdawg) [inline src/bot-handlers/disputer.js:89]: nit: could -> Could

File: src/bot-handlers/tests/disputer-test.js [hunk 8]
```diff
@@ -0,0 +1,526 @@
+/* eslint-disable no-loop-func, max-len */
```
Comment (arbitrary-kdawg) [inline src/bot-handlers/tests/disputer-test.js:1]: Would it be helpful to add a comment that indicates most of this code came from UMA source file xxx?

File: src/bot-handlers/disputer.js [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+            }
+            // get the historic price at the liquidation time.
+            let price;
+            try {
+              price = await priceFeed.getHistoricalPrice(liquidationTime);
+            } catch (error) {
+              console.error('Could not get historical price');
+            }
+            if (!price) return null;
+            // price is available, use it to determine if the liquidation is disputable
```
Comment (y0wl) [inline src/bot-handlers/disputer.js:95]: maybe moove the newline above the comment

File: hardhat.config.js [hunk 10]
```diff
@@ -0,0 +1,16 @@
+const { getHardhatConfig } = require("@uma/common");
+
+const path = require("path");
+const coreWkdir = path.dirname(require.resolve("@uma/contracts-node/package.json"));
+
+const configOverride = {
+  paths: {
+    root: coreWkdir,
+    //sources: `${coreWkdir}/contracts`,
```
[the remark this answers is not in the document]
Comment (y0wl) [reply hardhat.config.js]: this was removed already I belive

**Pull request merged**